posts/go-cli-from-scratch.md

```diff
@@ -0,0 +1,768 @@
+```
+
+Now you save the template using this subcommand:
+
+```shell
+$ go run main.go add --file test.md
+```
+
+We should now be able to save templates but how can we confirm it's actually saved. This is where the next command `list` comes into play.
+
+### Add List Command
+
+This command will list all the available templates in the template folder. With this, we can validate the `test.md` template was added successfully. Similar to the add command we'll be relying on the `os` package to manage files and folder paths.
+
+Jumping back to our steps, we'll first define the usage message:
+
+```go
+// list.go
+var listUsage = `List all currently avaible templates.
+
+Usage: brief list
+Options:
+`
+```
+
+Then create the new list command, this will be simpler than others since it does not use any specific flags.
+
+```go
+// list.go
+func NewListCommand() *Command {
+	cmd := &Command{
+		flags:   flag.NewFlagSet("list", flag.ExitOnError),
+		Execute: listFunc,
+	}
+	...
+```
+
+So this command needs to list all templates in the specific folder in the user home directory.
+
+We'll use `os.UserHomeDir()` to grab the user-specific home directory then look for the template folder in the app subdirectory.
+
+Finally, we use some basic string formatting for a friendly output.
+
+```go
+// list.go
+var listFunc = func(cmd *Command, args []string) {
+	homeDir, err := os.UserHomeDir()
+	... // file error handling
+
+	fileDir := filepath.Join(homeDir, ".gupi", "template")
+	file, err := os.Open(fileDir)
+	... // file error handling
+
+	filelist, err := file.Readdir(0)
+	... // file error handling
+
+	fmt.Printf("NAME\t\tSIZE\t\tMODIFIED")
+	for _, files := range filelist {
+		fmt.Printf("\n%-15s %-15v %v\n", files.Name(), files.Size(), files.ModTime().Format("2006-01-02 15:04:05"))
+	}
+}
+```
+
+The last step is to add another case in the switch statement
+
+```go
+// main.go
+case "list":
+	cmd = command.NewListCommand()
+```
+
+Now we can test this command by running:
+
+```shell
+$ go run main.go list
+NAME		SIZE		MODIFIED
+test.md     24          2021-12-02 00:44:10
+```
+
+The next command will allow us to edit any existing templates in terminal.
+
+### Add Edit Command
+
+This next command is fun to use and surprisingly simple to put together. We'll add the ability to edit our saved templates directly from the terminal using your editor of choice. I'll be using Vim as my editor but this could easily be modified into a flag or config variable to be passed in.
+
+The first thing to do is more familiar with error handling, we know where templates live and this command requires a template name so we just have to make sure it all exists as expected.
+
+Then we will define what command will be executed to start the editor, in this case, we will run `vim FILE_NAME`. Before we actually execute the command we need to configure the IO to be sent to our terminal. Along with that some simple error handling in case anything goes wrong along the way.
```

Comment:
```suggestion
Then we will define what command will be executed to start the editor. In this case, we will run `vim FILE_NAME`. Before we actually execute the command we need to configure the IO to be sent to our terminal. Along with that some simple error handling in case anything goes wrong along the way.
```